src/VisualStudio/Core/Def/Implementation/ProjectSystem/RunningDocumentTableEventTracker.cs

```diff
@@ -0,0 +1,257 @@
+            if ((grfAttribs & (uint)__VSRDTATTRIB.RDTA_DocDataReloaded) != 0)
+            {
+                _foregroundAffinitization.AssertIsForeground();
+                if (_runningDocumentTable.IsDocumentInitialized(docCookie) && TryGetMoniker(docCookie, out var moniker) && TryGetBuffer(docCookie, out var buffer))
+                {
+                    _runningDocumentTable.GetDocumentHierarchyItem(docCookie, out var hierarchy, out _);
+                    _listener.OnOpenDocument(moniker, buffer, hierarchy);
+                }
+            }
+
+            if ((grfAttribs & (uint)__VSRDTATTRIB.RDTA_Hierarchy) != 0)
+            {
+                _foregroundAffinitization.AssertIsForeground();
+                if (_runningDocumentTable.IsDocumentInitialized(docCookie) && TryGetMoniker(docCookie, out var moniker))
+                {
+                    _runningDocumentTable.GetDocumentHierarchyItem(docCookie, out var hierarchy, out _);
+                    _listener.OnRefreshDocumentContext(moniker, hierarchy);
+                }
+            }
+
+            return VSConstants.S_OK;
+        }
+
+        public int OnBeforeDocumentWindowShow(uint docCookie, int fFirstShow, IVsWindowFrame pFrame)
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public int OnAfterDocumentWindowHide(uint docCookie, IVsWindowFrame pFrame)
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public int OnBeforeSave(uint docCookie)
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public bool IsMonikerValid(string fileName) => _runningDocumentTable.IsMonikerValid(fileName);
+
+        /// <summary>
+        /// Attempts to get a text buffer from the specified moniker.
+        /// </summary>
+        /// <param name="moniker">the moniker to retrieve the text buffer for.</param>
+        /// <param name="textBuffer">the output text buffer or null if the moniker is invalid / document is not initialized.</param>
+        /// <returns>true if the buffer was found with a non null value.</returns>
+        public bool TryGetBufferFromMoniker(string moniker, out ITextBuffer textBuffer)
+        {
+            _foregroundAffinitization.AssertIsForeground();
+
+            textBuffer = null;
+            if (!IsMonikerValid(moniker))
+            {
+                return false;
+            }
+
+            var cookie = _runningDocumentTable.GetDocumentCookie(moniker);
+            if (!_runningDocumentTable.IsDocumentInitialized(cookie))
+            {
+                return false;
+            }
+
+            return TryGetBuffer(cookie, out textBuffer);
+        }
+
+        /// <summary>
+        /// Applies an action to all initialized files in the RDT.
+        /// </summary>
+        /// <param name="enumerateAction">the action to apply.</param>
+        public void EnumerateDocumentSet(Action<string, ITextBuffer, IVsHierarchy> enumerateAction)
```

Comment:
The reason I did it this way was to avoid having to return all three pieces of data.  But an IEnumerable of tuple seems like a reasonable way to do it.  Will update.